sound/soc/sof/intel/hda.c

```diff
@@ -1021,32 +1021,65 @@ static bool link_slaves_found(struct snd_sof_dev *sdev,
 	struct sdw_intel_slave_id *ids = sdw->ids;
 	int num_slaves = sdw->num_slaves;
 	unsigned int part_id, link_id, unique_id, mfg_id;
-	int i, j;
+	int i, j, k;
 
 	for (i = 0; i < link->num_adr; i++) {
 		u64 adr = link->adr_d[i].adr;
+		int reported_part_count = 0;
 
 		mfg_id = SDW_MFG_ID(adr);
 		part_id = SDW_PART_ID(adr);
 		link_id = SDW_DISCO_LINK_ID(adr);
+
+		for (j = 0; j < num_slaves; j++) {
+			/* find out how many identical parts were reported on that link */
+			if (ids[j].link_id == link_id &&
+			    ids[j].id.part_id == part_id &&
+			    ids[j].id.mfg_id == mfg_id)
+				reported_part_count++;
+		}
+
 		for (j = 0; j < num_slaves; j++) {
+			int expected_part_count = 0;
+
 			if (ids[j].link_id != link_id ||
 			    ids[j].id.part_id != part_id ||
 			    ids[j].id.mfg_id != mfg_id)
 				continue;
-			/*
-			 * we have to check unique id
-			 * if there is more than one
-			 * Slave on the link
-			 */
-			unique_id = SDW_UNIQUE_ID(adr);
-			if (link->num_adr == 1 ||
-			    ids[j].id.unique_id == SDW_IGNORED_UNIQUE_ID ||
-			    ids[j].id.unique_id == unique_id) {
-				dev_dbg(bus->dev,
-					"found %x at link %d\n",
-					part_id, link_id);
-				break;
+
+			/* find out how many identical parts are expected */
+			for (k = 0; k < link->num_adr; k++) {
+				u64 adr2 = link->adr_d[i].adr;
+				unsigned int part_id2, link_id2, mfg_id2;
+
+				mfg_id2 = SDW_MFG_ID(adr2);
+				part_id2 = SDW_PART_ID(adr2);
+				link_id2 = SDW_DISCO_LINK_ID(adr2);
+
+				if (link_id2 == link_id &&
+				    part_id2 == part_id &&
+				    mfg_id2 == mfg_id)
+					expected_part_count++;
+			}
+
+			if (reported_part_count == expected_part_count) {
+				/*
+				 * we have to check unique id
+				 * if there is more than one
+				 * Slave on the link
+				 */
+				unique_id = SDW_UNIQUE_ID(adr);
+				if (link->num_adr == 1 ||
+				    ids[j].id.unique_id == SDW_IGNORED_UNIQUE_ID ||
```

Comment:
I think we can replace `link->num_adr == 1 || ids[j].id.unique_id == SDW_IGNORED_UNIQUE_ID ||` with `reported_part_count == 1`. We should check unique id when there are more than one codecs with the same mgf_id and part_id.

Reply:
I am not sure we can replace both conditions, you could have link->num_adr > 1 in case we have e.g. two different amplifiers on the same link. It's never happened so far but at some point some hardware genius will enable this configuration.

But yes I think ids[j].id.unique_id == SDW_IGNORED_UNIQUE_ID and reported_part_count == 1 are identical.

Reply:
fixed now

Reply:
> I am not sure we can replace both conditions, you could have link->num_adr > 1 in case we have e.g. two different amplifiers on the same link. It's never happened so far but at some point some hardware genius will enable this configuration.

We will check the next condition (`reported_part_count == 1`) if `link->num_adr > 1` and `reported_part_count == 1` will be true if `link->num_adr == 1`. So I think it is fine to remvoe the `link->num_adr == 1` condition.

Reply:
ok, done. thanks @bardliao